<a href="https://colab.research.google.com/github/khairulniza/MANB2153/blob/master/Quiz2_Khairulniza.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
from google.colab import drive
drive.mount('/content/gdrive')


Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=email%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdocs.test%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.photos.readonly%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/gdrive


In [5]:
#read data from gd using pandas
import pandas as pd
df=pd.read_csv('gdrive/My Drive/heart.csv')
df.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

Q.1: It is a classification problem where we want to predict presense of heart disease in patient 

In [0]:
from sklearn.model_selection import train_test_split
X = df.drop("target", axis=1).values
y = df.iloc[:,-1:].values

In [0]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.30, random_state=42)

In [0]:
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical
#one-hot encode target column, heart issue put as [0 1] no issue [1 0] 
y_train = to_categorical(df.target)

#vcheck that target column has been converted
model=Sequential()
n_cols=X_train.shape[1]

In [0]:
model.add(Dense(250, activation='relu', input_shape=(n_cols,)))
model.add(Dense(250, activation='relu'))
model.add(Dense(250, activation='relu'))
model.add(Dense(250, activation='relu'))
model.add(Dense(1, activation='softmax'))

In [0]:
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [14]:
from keras.callbacks import EarlyStopping
early_stopping_monitor = EarlyStopping(patience=3)

model.fit(X, y, epochs=30, validation_split=0.2)

Train on 242 samples, validate on 61 samples
Epoch 1/30
242/242 [==============================] - 1s 2ms/step - loss: 5.0726 - acc: 0.6818 - val_loss: 15.9424 - val_acc: 0.0000e+00
Epoch 2/30
242/242 [==============================] - 0s 268us/step - loss: 5.0726 - acc: 0.6818 - val_loss: 15.9424 - val_acc: 0.0000e+00
Epoch 3/30
242/242 [==============================] - 0s 244us/step - loss: 5.0726 - acc: 0.6818 - val_loss: 15.9424 - val_acc: 0.0000e+00
Epoch 4/30
242/242 [==============================] - 0s 254us/step - loss: 5.0726 - acc: 0.6818 - val_loss: 15.9424 - val_acc: 0.0000e+00
Epoch 5/30
242/242 [==============================] - 0s 278us/step - loss: 5.0726 - acc: 0.6818 - val_loss: 15.9424 - val_acc: 0.0000e+00
Epoch 6/30
242/242 [==============================] - 0s 265us/step - loss: 5.0726 - acc: 0.6818 - val_loss: 15.9424 - val_acc: 0.0000e+00
Epoch 7/30
242/242 [==============================] - 0s 282us/step - loss: 5.0726 - acc: 0.6818 - val_loss: 15.9424 - val_

Q. 3.e Accuray after 20 epoch is 68.18%

Q.4 Overfitting refers to a model that models the training data too well. when a model learns the detail and noise in the training data to the extent that it negatively impacts the performance of the model on new data

Underfitting refers to a model that can neither model the training data nor generalize to new data. The machine learning model is not a suitable model and will be obvious as it will have poor performance on the training data

Q.5 To reduce overfitting is to lower the capacity of the model to memorize the training data. As such, the model will need to focus on the relevant patterns in the training data, which results in better generalization. In this post, we’ll discuss three options to achieve this